src/main/java/server/request/Parser.java

```diff
@@ -23,23 +26,20 @@ public Request parse() {
         return new Request(requestMethod, requestPath, requestHeaders, requestBody);
     }
 
-    private String parseRequestLine() {
-        return client.read();
+    private String parseRequestLine() throws IOException {
+        String requestLine = client.read();
+        return requestLine == null ? "" : requestLine;
```

Comment:
In general I prefer to be explicit about failures like this, rather than return a value that allows an odd state to exist. In other words, from the interface level it's not overly clear that this may return an empty string, or what that even means in this context. Eventually that empty string makes it's way into the Selector after some additional parsing, but that's pretty far from the source. 

One option would be to use `java.util.Optional`s to be explicit about the lack of real value here. While it would be similar to returning null, it allows the caller to optionally act on the value if it is present, otherwise it becomes a NOOP. I started playing around with this, but it doesn't quite fit into the current handler/selector structure. I'm fine with this for now, just wanted to call that out. 

Outside of this specific feedback I'd recommend looking into Optionals if you haven't already, they can be very powerful, and in my experience the readability of the code is quite nice compared to null/empty checks all over.

Reply:
To me it made sense to just assign an empty string to a `Request` field in the case that the request is empty and use that value to determine what kind of response to send--do you think the solution here could be as simple as introducing a meaningfully named temp variable and return something like `emptyRequestLine` to clarify what's happening?

Reply:
That's fair. I guess I was wondering about reducing the scope in which that empty string context exists. With your explanation I think it makes sense in the context of the Parser, but maybe less so in the Selector. I'm fine with this for now.